### ISSUE TITLE
"Strategy Pattern" makes static inference impossible without type hints and leads to inferior results compared to IPython's old auto complete.

### ISSUE THREAD

**Issue by fmagin** [davidhalter/jedi#1453]
## Context
I am not that familiar with jedi internals and this is my first issue that led me to dive into them. I tried debugging this issue because I initially thought it was related to https://github.com/davidhalter/jedi/issues/1058 but it turned out that jedi fails earlier already.

The angr project uses a plugin system that is based on the strategy pattern according to the documentation. The general idea and issue with jedi is the following:

Some central object like an instance of [`AnalysisHub`](https://github.com/angr/angr/blob/05cb4839c671f90973543e01a7b7b58cddbb6d71/angr/analyses/analysis.py#L56) or anything that inherits from [`PluginHub`](https://github.com/angr/angr/blob/2b3ee30c096465f8992f926eaf6d06818aeb578b/angr/misc/plugins.py#L9) exposes a method that allows registering a class with `register_plugin(some_cls, 'some_name')` and then using `proj.analyses.some_name(some_keyword_arg='kw')` where `analyses` is the instance of `AnalysisHub`. The object returned by accessing `some_name` is an [`AnalysisFactory`](https://github.com/angr/angr/blob/05cb4839c671f90973543e01a7b7b58cddbb6d71/angr/analyses/analysis.py#L80) which's `__call__` implementation wraps the `__init__` of `some_cls`. The final expectation is that IPython can auto complete the keyword arguments inside this call. This works with the old autocompleter of IPython but not with jedi. As far as I understood the code while debugging this the issue is the following:

## Bug
Jedi gets as far as correctly identifying `proj.analyses` as an `AnalysisHub` but then fails to infer the type of `proj.analyses.some_name` and hence the signature can never be generated.

Assumption: The old auto completion probably just accesses the actual object directly and finds `__signature__` of `__call__` of the object and uses it. Jedi tries some kind of static approach that fails because it is impossible to find out the type of `proj.analyses.some_name` from just knowing the type of `proj.analyses`.

Suggestion: Some kind of fallback or additional inference that also checks if the actual object has the information for tab completions available even if it is impossible to statically find out why this is the case.

Due to @davidhalter's https://github.com/davidhalter/jedi/issues/1058#issuecomment-547685948
```
I have no idea what your issue could be, but as a first indicator, please provide inspect.signature(...) of your object. If that works, Jedi should work without issues.
```
I am considering this a bug.
Together with some other issues this leads me to actively disable jedi in IPython for me, which I rather want to avoid in the long term.
I can understand though that this is only possible in the context of interactive use and the underlying code pattern makes this impossible to deal with in an IDE, unless using type annotations on fake class variables. I haven't managed to get it working even if type annotations were present, but I did not properly investigate this. If that would work I would consider this a far less severe issue because in those cases type annotations are highly useful anyway.

If there is anything I can do to track this down I am happy to debug things and provide further context and information, my main issue is that I do not have a high level overview of how jedi is generating the completions and cannot tell at which point exactly it starts to fail.

## Setup to reproduce

Install angr in a virtualenv, `pip install angr` should be enough.

Run the following script in IPython:
```python
import inspect
import angr
proj = angr.Project('/bin/true', load_options={'auto_load_libs': False})
```

Confirm that `inspect` works:
```python
In [3]: inspect.signature(proj.analyses.CFGFast)                                                          
Out[3]: <Signature (binary=None, objects=None, regions=None, pickle_intermediate_results=False, symbols=True, function_prologues=True, resolve_indirect_jumps=True, force_segment=False, force_complete_scan=True, indirect_jump_target_limit=100000, data_references=False, cross_references=False, normalize=False, start_at_entry=True, function_starts=None, extra_memory_regions=None, data_type_guessing_handlers=None, arch_options=None, indirect_jump_resolvers=None, base_state=None, exclude_sparse_regions=True, skip_specific_regions=True, heuristic_plt_resolving=None, detect_tail_calls=False, low_priority=False, cfb=None, model=None, use_patches=False, start=None, end=None, collect_data_references=None, extra_cross_references=None, **extra_arch_options)>
```

Confirm that tab completion for those arguments fails:
```python
proj.analyses.CFGFast(for<tab>
```

**Comment by davidhalter** [comment]
Since the tab completion there involves quite a few things, I would be interested what something like this outputs:

```
import jedi
import inspect
import angr
jedi.set_debug_function()
proj = angr.Project('/bin/true', load_options={'auto_load_libs': False})
print(jedi.Interpreter('proj.analyses.CFGFast(', [globals()] ).call_signatures())
print(inspect.signature(proj.analyses.CFGFast))
```

And if that's possible, it would also be interesting if you can create a simpler case for `proj`, where this variable is defined with simple Python stuff.

**Comment by fmagin** [comment]
Versions:
Jedi: 8e33fd1931dc21619ec81f5b7a1e9d0f59248dfe
Parso: e496b07b6342f6182225a60aad6031d7ad08f24d

```
In [1]: import jedi 
   ...: import inspect 
   ...: import angr 
   ...: jedi.set_debug_function() 
   ...: proj = angr.Project('/bin/true', load_options={'auto_load_libs': False})                 
/home/fmagin/.virtualenvs/jedi/lib/python3.8/site-packages/pysmt/walkers/generic.py:43: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
  if len(nodetypes) == 1 and isinstance(nodetypes[0], collections.Iterable):
WARNING | 2019-12-09 16:06:48,683 | cle.loader | The main binary is a position-independent executable. It is being loaded with a base address of 0x400000.
```

```
In [2]: print(jedi.Interpreter('proj.analyses.CFGFast(', [globals()] ).call_signatures())        
speed: init 63.35204339027405
speed: parsed 63.35867214202881
 dbg: infer_node <Name: proj@1,0>@(1, 0) in MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)
 dbg: context.goto <Name: proj@1,0> in (MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)): [<MixedName: (<CompiledValueName: string_name=>).proj>]
 dbg: global search_module 'builtins': <CompiledObject: <module 'builtins' (built-in)>>
 dbg: execute: <ClassValue: <Class: Project@38-708>> <ValuesArguments: []>
 dbg: execute result: S{<TreeInstance of <ClassValue: <Class: Project@38-708>>(<ValuesArguments: []>)>} in <ClassValue: <Class: Project@38-708>>
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=>).proj>] -> S{<MixedObject: <angr.project.Project object at 0x7f13354cd6d0>>}
dbg: infer_trailer: PythonNode(trailer, [<Operator: .>, <Name: analyses@1,5>]) in S{<MixedObject: <angr.project.Project object at 0x7f13354cd6d0>>}
dbg: context.goto <Name: analyses@1,5> in (<MixedObject: <angr.project.Project object at 0x7f13354cd6d0>>): [<MixedName: (<ValueName: string_name=Project start_pos=(38, 6)>).analyses>]
dbg: execute: <ClassValue: <Class: AnalysesHub@56-78>> <ValuesArguments: []>
dbg: execute result: S{<TreeInstance of <ClassValue: <Class: AnalysesHub@56-78>>(<ValuesArguments: []>)>} in <ClassValue: <Class: AnalysesHub@56-78>>
dbg: context.names_to_types: [<MixedName: (<ValueName: string_name=Project start_pos=(38, 6)>).analyses>] -> S{<MixedObject: <angr.analyses.analysis.AnalysesHub object at 0x7f13357b67f0>>}
dbg: infer_trailer: PythonNode(trailer, [<Operator: .>, <Name: CFGFast@1,14>]) in S{<MixedObject: <angr.analyses.analysis.AnalysesHub object at 0x7f13357b67f0>>}
dbg: context.goto <Name: CFGFast@1,14> in (<MixedObject: <angr.analyses.analysis.AnalysesHub object at 0x7f13357b67f0>>): []
 dbg: infer_node <Name: PluginVendor@56,18>@(56, 18) in MixedModuleContext(<MixedObject: <module 'angr.analyses.analysis' from '/home/fmagin/Projects/angr-dev/angr/angr/analyses/analysis.py'>>)
 dbg: context.goto <Name: PluginVendor@56,18> in (MixedModuleContext(<MixedObject: <module 'angr.analyses.analysis' from '/home/fmagin/Projects/angr-dev/angr/angr/analyses/analysis.py'>>)): [<MixedName: (<CompiledValueName: string_name=analysis>).PluginVendor>]
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=analysis>).PluginVendor>] -> S{<MixedObject: <class 'angr.misc.plugins.PluginVendor'>>}
 dbg: infer_node <Name: PluginHub@261,19>@(261, 19) in MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)
 dbg: context.goto <Name: PluginHub@261,19> in (MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)): [<MixedName: (<CompiledValueName: string_name=plugins>).PluginHub>]
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=plugins>).PluginHub>] -> S{<MixedObject: <class 'angr.misc.plugins.PluginHub'>>}
 dbg: infer_node <Name: object@9,16>@(9, 16) in MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)
 dbg: context.goto <Name: object@9,16> in (MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)): [<_StubName: string_name=object start_pos=(40, 6)>]
 dbg: context.names_to_types: [<_StubName: string_name=object start_pos=(40, 6)>] -> S{<ClassValue: <Class: object@40-70>>}
dbg: context.goto '__init__' in (<ClassValue: <Class: AnalysesHub@56-78>>): [<ClassName: string_name=__init__ start_pos=(60, 8)>]
dbg: context.names_to_types: [<ClassName: string_name=__init__ start_pos=(60, 8)>] -> S{<MethodValue: <Function: __init__@60-63>>}
dbg: Overloading match: '__init__(self, project)'@60 (<InstanceArguments: <ValuesArguments: []>>)
dbg: execute: <BoundMethod: <MethodValue: <Function: __getattr__@71-76>>> <ValuesArguments: [S{<CompiledValue: <CompiledObject: 'CFGFast'>>}]>
  dbg: infer_node <String: 'SimStatePlugin'>@(71, 40) in ClassContext(<ClassValue: <Class: PluginHub@9-203>>)
  dbg: infer_node PythonNode(eval_input, [<Name: SimStatePlugin@289,0>, <EndMarker: prefix='' end_pos=(289, 14)>])@(289, 0) in ClassContext(<ClassValue: <Class: PluginHub@9-203>>)
   dbg: infer_node <Name: SimStatePlugin@289,0>@(289, 0) in ClassContext(<ClassValue: <Class: PluginHub@9-203>>)
   dbg: context.goto <Name: SimStatePlugin@289,0> in (ClassContext(<ClassValue: <Class: PluginHub@9-203>>)): [<MixedName: (<CompiledValueName: string_name=plugins>).SimStatePlugin>]
   dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=plugins>).SimStatePlugin>] -> S{<MixedObject: <class 'angr.state_plugins.plugin.SimStatePlugin'>>}
 dbg: execute: <MixedObject: <class 'angr.state_plugins.plugin.SimStatePlugin'>> <ValuesArguments: []>
 dbg: execute result: S{<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>} in <MixedObject: <class 'angr.state_plugins.plugin.SimStatePlugin'>>
dbg: execute result: S{<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>} in <BoundMethod: <MethodValue: <Function: __getattr__@71-76>>>
dbg: context.names_to_types: [] -> S{<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>}
speed: func_call followed 0.46700000762939453
 dbg: infer_node <Name: object@10,21>@(10, 21) in MixedModuleContext(<MixedObject: <module 'angr.state_plugins.plugin' from '/home/fmagin/Projects/angr-dev/angr/angr/state_plugins/plugin.py'>>)
 dbg: context.goto <Name: object@10,21> in (MixedModuleContext(<MixedObject: <module 'angr.state_plugins.plugin' from '/home/fmagin/Projects/angr-dev/angr/angr/state_plugins/plugin.py'>>)): [<_StubName: string_name=object start_pos=(40, 6)>]
 dbg: context.names_to_types: [<_StubName: string_name=object start_pos=(40, 6)>] -> S{<ClassValue: <Class: object@40-70>>}
dbg: context.goto '__init__' in (<ClassValue: <Class: SimStatePlugin@10-130>>): [<ClassName: string_name=__init__ start_pos=(19, 8)>]
dbg: context.names_to_types: [<ClassName: string_name=__init__ start_pos=(19, 8)>] -> S{<MethodValue: <Function: __init__@19-21>>}
dbg: Overloading match: '__init__(self)'@19 (<InstanceArguments: <ValuesArguments: []>>)
dbg: context.goto '__call__' in (<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>): []
dbg: execute: <BoundMethod: <MethodValue: <Function: __getattribute__@61-62>>> <ValuesArguments: [S{<CompiledValue: <CompiledObject: '__call__'>>}]>
  dbg: infer_node <Name: Any@61,45>@(61, 45) in ClassContext(<ClassValue: <Class: object@40-70>>)
  dbg: context.goto <Name: Any@61,45> in (ClassContext(<ClassValue: <Class: object@40-70>>)): [<TreeNameDefinition: string_name=Any start_pos=(6, 69)>]
  speed: import (<Name: typing@4,5>,) ModuleContext(<StubModuleValue: builtins@4-1651 is_stub=True>) 0.4694674015045166
  dbg: global search_module 'typing': <ModuleValue: typing@1-1990 is_stub=False>
  dbg: context.goto <Name: Any@6,69> in (<TypingModuleWrapper: typing@3-646 is_stub=True>): [TypingModuleName(<_StubName: string_name=Any start_pos=(12, 0)>)]
  dbg: context.names_to_types: [TypingModuleName(<_StubName: string_name=Any start_pos=(12, 0)>)] -> S{<Any>}
  dbg: after import: S{<Any>}
  dbg: context.names_to_types: [<TreeNameDefinition: string_name=Any start_pos=(6, 69)>] -> S{<Any>}
 dbg: Start: Resolve lazy value wrapper
 dbg: End: Resolve lazy value wrapper
 warning: Used Any - returned no results
dbg: execute result: S{} in <BoundMethod: <MethodValue: <Function: __getattribute__@61-62>>>
dbg: context.names_to_types: [] -> S{}
[]
```

```
In [3]: print(inspect.signature(proj.analyses.CFGFast))                                          
(binary=None, objects=None, regions=None, pickle_intermediate_results=False, symbols=True, function_prologues=True, resolve_indirect_jumps=True, force_segment=False, force_complete_scan=True, indirect_jump_target_limit=100000, data_references=False, cross_references=False, normalize=False, start_at_entry=True, function_starts=None, extra_memory_regions=None, data_type_guessing_handlers=None, arch_options=None, indirect_jump_resolvers=None, base_state=None, exclude_sparse_regions=True, skip_specific_regions=True, heuristic_plt_resolving=None, detect_tail_calls=False, low_priority=False, cfb=None, model=None, use_patches=False, start=None, end=None, collect_data_references=None, extra_cross_references=None, **extra_arch_options)
```

I am not sure what you mean by 

> a simpler case for proj, where this variable is defined with simple Python stuff.

`proj` shouldn't be that relevant, it still fails with:

```
analyses = proj.analyses
print(jedi.Interpreter('analyses.CFGFast(', [globals()] ).call_signatures())             
speed: init 30.10172963142395
speed: parsed 30.108437538146973
 dbg: infer_node <Name: analyses@1,0>@(1, 0) in MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)
 dbg: context.goto <Name: analyses@1,0> in (MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)): [<MixedName: (<CompiledValueName: string_name=>).analyses>]
 dbg: global search_module 'builtins': <CompiledObject: <module 'builtins' (built-in)>>
 dbg: execute: <ClassValue: <Class: AnalysesHub@56-78>> <ValuesArguments: []>
 dbg: execute result: S{<TreeInstance of <ClassValue: <Class: AnalysesHub@56-78>>(<ValuesArguments: []>)>} in <ClassValue: <Class: AnalysesHub@56-78>>
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=>).analyses>] -> S{<MixedObject: <angr.analyses.analysis.AnalysesHub object at 0x7fb18eaa70a0>>}
dbg: infer_trailer: PythonNode(trailer, [<Operator: .>, <Name: CFGFast@1,9>]) in S{<MixedObject: <angr.analyses.analysis.AnalysesHub object at 0x7fb18eaa70a0>>}
dbg: context.goto <Name: CFGFast@1,9> in (<MixedObject: <angr.analyses.analysis.AnalysesHub object at 0x7fb18eaa70a0>>): []
 dbg: infer_node <Name: PluginVendor@56,18>@(56, 18) in MixedModuleContext(<MixedObject: <module 'angr.analyses.analysis' from '/home/fmagin/Projects/angr-dev/angr/angr/analyses/analysis.py'>>)
 dbg: context.goto <Name: PluginVendor@56,18> in (MixedModuleContext(<MixedObject: <module 'angr.analyses.analysis' from '/home/fmagin/Projects/angr-dev/angr/angr/analyses/analysis.py'>>)): [<MixedName: (<CompiledValueName: string_name=analysis>).PluginVendor>]
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=analysis>).PluginVendor>] -> S{<MixedObject: <class 'angr.misc.plugins.PluginVendor'>>}
 dbg: infer_node <Name: PluginHub@261,19>@(261, 19) in MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)
 dbg: context.goto <Name: PluginHub@261,19> in (MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)): [<MixedName: (<CompiledValueName: string_name=plugins>).PluginHub>]
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=plugins>).PluginHub>] -> S{<MixedObject: <class 'angr.misc.plugins.PluginHub'>>}
 dbg: infer_node <Name: object@9,16>@(9, 16) in MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)
 dbg: context.goto <Name: object@9,16> in (MixedModuleContext(<MixedObject: <module 'angr.misc.plugins' from '/home/fmagin/Projects/angr-dev/angr/angr/misc/plugins.py'>>)): [<_StubName: string_name=object start_pos=(40, 6)>]
 dbg: context.names_to_types: [<_StubName: string_name=object start_pos=(40, 6)>] -> S{<ClassValue: <Class: object@40-70>>}
dbg: context.goto '__init__' in (<ClassValue: <Class: AnalysesHub@56-78>>): [<ClassName: string_name=__init__ start_pos=(60, 8)>]
dbg: context.names_to_types: [<ClassName: string_name=__init__ start_pos=(60, 8)>] -> S{<MethodValue: <Function: __init__@60-63>>}
dbg: Overloading match: '__init__(self, project)'@60 (<InstanceArguments: <ValuesArguments: []>>)
dbg: execute: <BoundMethod: <MethodValue: <Function: __getattr__@71-76>>> <ValuesArguments: [S{<CompiledValue: <CompiledObject: 'CFGFast'>>}]>
  dbg: infer_node <String: 'SimStatePlugin'>@(71, 40) in ClassContext(<ClassValue: <Class: PluginHub@9-203>>)
  dbg: infer_node PythonNode(eval_input, [<Name: SimStatePlugin@289,0>, <EndMarker: prefix='' end_pos=(289, 14)>])@(289, 0) in ClassContext(<ClassValue: <Class: PluginHub@9-203>>)
   dbg: infer_node <Name: SimStatePlugin@289,0>@(289, 0) in ClassContext(<ClassValue: <Class: PluginHub@9-203>>)
   dbg: context.goto <Name: SimStatePlugin@289,0> in (ClassContext(<ClassValue: <Class: PluginHub@9-203>>)): [<MixedName: (<CompiledValueName: string_name=plugins>).SimStatePlugin>]
   dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=plugins>).SimStatePlugin>] -> S{<MixedObject: <class 'angr.state_plugins.plugin.SimStatePlugin'>>}
 dbg: execute: <MixedObject: <class 'angr.state_plugins.plugin.SimStatePlugin'>> <ValuesArguments: []>
 dbg: execute result: S{<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>} in <MixedObject: <class 'angr.state_plugins.plugin.SimStatePlugin'>>
dbg: execute result: S{<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>} in <BoundMethod: <MethodValue: <Function: __getattr__@71-76>>>
dbg: context.names_to_types: [] -> S{<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>}
speed: func_call followed 0.07377338409423828
 dbg: infer_node <Name: object@10,21>@(10, 21) in MixedModuleContext(<MixedObject: <module 'angr.state_plugins.plugin' from '/home/fmagin/Projects/angr-dev/angr/angr/state_plugins/plugin.py'>>)
 dbg: context.goto <Name: object@10,21> in (MixedModuleContext(<MixedObject: <module 'angr.state_plugins.plugin' from '/home/fmagin/Projects/angr-dev/angr/angr/state_plugins/plugin.py'>>)): [<_StubName: string_name=object start_pos=(40, 6)>]
 dbg: context.names_to_types: [<_StubName: string_name=object start_pos=(40, 6)>] -> S{<ClassValue: <Class: object@40-70>>}
dbg: context.goto '__init__' in (<ClassValue: <Class: SimStatePlugin@10-130>>): [<ClassName: string_name=__init__ start_pos=(19, 8)>]
dbg: context.names_to_types: [<ClassName: string_name=__init__ start_pos=(19, 8)>] -> S{<MethodValue: <Function: __init__@19-21>>}
dbg: Overloading match: '__init__(self)'@19 (<InstanceArguments: <ValuesArguments: []>>)
dbg: context.goto '__call__' in (<TreeInstance of <ClassValue: <Class: SimStatePlugin@10-130>>(<ValuesArguments: []>)>): []
dbg: execute: <BoundMethod: <MethodValue: <Function: __getattribute__@61-62>>> <ValuesArguments: [S{<CompiledValue: <CompiledObject: '__call__'>>}]>
  dbg: infer_node <Name: Any@61,45>@(61, 45) in ClassContext(<ClassValue: <Class: object@40-70>>)
  dbg: context.goto <Name: Any@61,45> in (ClassContext(<ClassValue: <Class: object@40-70>>)): [<TreeNameDefinition: string_name=Any start_pos=(6, 69)>]
  speed: import (<Name: typing@4,5>,) ModuleContext(<StubModuleValue: builtins@4-1651 is_stub=True>) 0.07633829116821289
  dbg: global search_module 'typing': <ModuleValue: typing@1-1990 is_stub=False>
  dbg: context.goto <Name: Any@6,69> in (<TypingModuleWrapper: typing@3-646 is_stub=True>): [TypingModuleName(<_StubName: string_name=Any start_pos=(12, 0)>)]
  dbg: context.names_to_types: [TypingModuleName(<_StubName: string_name=Any start_pos=(12, 0)>)] -> S{<Any>}
  dbg: after import: S{<Any>}
  dbg: context.names_to_types: [<TreeNameDefinition: string_name=Any start_pos=(6, 69)>] -> S{<Any>}
 dbg: Start: Resolve lazy value wrapper
 dbg: End: Resolve lazy value wrapper
 warning: Used Any - returned no results
dbg: execute result: S{} in <BoundMethod: <MethodValue: <Function: __getattribute__@61-62>>>
dbg: context.names_to_types: [] -> S{}
[]
```

but works with:

```
In [4]: CFGFast = proj.analyses.CFGFast                                                          
In [5]: print(jedi.Interpreter('CFGFast(', [globals()] ).call_signatures())                      
speed: init 96.9371247291565
speed: parsed 96.93754267692566
 dbg: infer_node <Name: CFGFast@1,0>@(1, 0) in MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)
 dbg: context.goto <Name: CFGFast@1,0> in (MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)): [<MixedName: (<CompiledValueName: string_name=>).CFGFast>]
 dbg: global search_module 'builtins': <CompiledObject: <module 'builtins' (built-in)>>
 dbg: execute: <ClassValue: <Class: AnalysisFactory@80-113>> <ValuesArguments: []>
 dbg: execute result: S{<TreeInstance of <ClassValue: <Class: AnalysisFactory@80-113>>(<ValuesArguments: []>)>} in <ClassValue: <Class: AnalysisFactory@80-113>>
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=>).CFGFast>] -> S{<MixedObject: <angr.analyses.analysis.AnalysisFactory object at 0x7fb18ea04460>>}
speed: func_call followed 0.0013952255249023438
[<CallSignature: index=0 AnalysisFactory(binary=None, objects=None, regions=None, pickle_intermediate_results=False, symbols=True, function_prologues=True, resolve_indirect_jumps=True, force_segment=False, force_complete_scan=True, indirect_jump_target_limit=100000, data_references=False, cross_references=False, normalize=False, start_at_entry=True, function_starts=None, extra_memory_regions=None, data_type_guessing_handlers=None, arch_options=None, indirect_jump_resolvers=None, base_state=None, exclude_sparse_regions=True, skip_specific_regions=True, heuristic_plt_resolving=None, detect_tail_calls=False, low_priority=False, cfb=None, model=None, use_patches=False, start=None, end=None, collect_data_references=None, extra_cross_references=None, **extra_arch_options)>]
```

When assigning CFGFast to some object it also works:
```
In [6]: proj.CFGFast = proj.analyses.CFGFast
In [8]: print(jedi.Interpreter('proj.CFGFast(', [globals()] ).call_signatures())                 
speed: init 5.654462814331055
speed: parsed 5.6548168659210205
 dbg: infer_node <Name: proj@1,0>@(1, 0) in MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)
 dbg: context.goto <Name: proj@1,0> in (MixedModuleContext(<ModuleValue: @1-1 is_stub=False>)): [<MixedName: (<CompiledValueName: string_name=>).proj>]
 dbg: global search_module 'builtins': <CompiledObject: <module 'builtins' (built-in)>>
 dbg: execute: <ClassValue: <Class: Project@38-708>> <ValuesArguments: []>
 dbg: execute result: S{<TreeInstance of <ClassValue: <Class: Project@38-708>>(<ValuesArguments: []>)>} in <ClassValue: <Class: Project@38-708>>
 dbg: context.names_to_types: [<MixedName: (<CompiledValueName: string_name=>).proj>] -> S{<MixedObject: <angr.project.Project object at 0x7fb1b84683a0>>}
dbg: infer_trailer: PythonNode(trailer, [<Operator: .>, <Name: CFGFast@1,5>]) in S{<MixedObject: <angr.project.Project object at 0x7fb1b84683a0>>}
dbg: context.goto <Name: CFGFast@1,5> in (<MixedObject: <angr.project.Project object at 0x7fb1b84683a0>>): [<MixedName: (<ValueName: string_name=Project start_pos=(38, 6)>).CFGFast>]
dbg: execute: <ClassValue: <Class: AnalysisFactory@80-113>> <ValuesArguments: []>
dbg: execute result: S{<TreeInstance of <ClassValue: <Class: AnalysisFactory@80-113>>(<ValuesArguments: []>)>} in <ClassValue: <Class: AnalysisFactory@80-113>>
dbg: context.names_to_types: [<MixedName: (<ValueName: string_name=Project start_pos=(38, 6)>).CFGFast>] -> S{<MixedObject: <angr.analyses.analysis.AnalysisFactory object at 0x7fb18ea04460>>}
speed: func_call followed 0.01474618911743164
[<CallSignature: index=0 AnalysisFactory(binary=None, objects=None, regions=None, pickle_intermediate_results=False, symbols=True, function_prologues=True, resolve_indirect_jumps=True, force_segment=False, force_complete_scan=True, indirect_jump_target_limit=100000, data_references=False, cross_references=False, normalize=False, start_at_entry=True, function_starts=None, extra_memory_regions=None, data_type_guessing_handlers=None, arch_options=None, indirect_jump_resolvers=None, base_state=None, exclude_sparse_regions=True, skip_specific_regions=True, heuristic_plt_resolving=None, detect_tail_calls=False, low_priority=False, cfb=None, model=None, use_patches=False, start=None, end=None, collect_data_references=None, extra_cross_references=None, **extra_arch_options)>]
```

Happy to provide further help and results and answer any question about angr internals that might be relevant.

**Comment by davidhalter** [comment]
Since I just fixed #1378, I will close this. I'm pretty sure it's the same issue. Otherwise let me know and I will reopen. If you want to test it you have to use both Jedi and Parso master branch.

Thanks again for the well written issue. Really helps with fixing these things.
